src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java

```diff
@@ -0,0 +1,247 @@
+        createUpstreamJobAndRun();
+
+        j.assertBuildStatus(FAILURE, upstreamRun);
+        j.assertLogContains(format("Physical path on disk was not provided in the Jenkins global config for Disk ID: '%s', within Disk Pool ID '%s'", DISK_ID_ONE, DISK_POOL_ID), upstreamRun);
+    }
+
+    @Test
+    public void physicalPathOnDiskNotRelative() throws Exception {
+        setUpDiskPool(new Disk(DISK_ID_ONE, "name", "mount", "/path"));
+        createUpstreamJobAndRun();
+
+        j.assertBuildStatus(FAILURE, upstreamRun);
+        j.assertLogContains(format("Physical path on disk defined for Disk ID '%s', within Disk Pool ID '%s' must be a relative path", DISK_ID_ONE, DISK_POOL_ID), upstreamRun);
+    }
+
+    @Test
+    public void successfullyAllocateWorkspace() throws Exception {
+        Disk disk1 = new Disk(DISK_ID_ONE, "name", pathToDisk1.getPath(), "path");
+        Disk disk2 = new Disk(DISK_ID_TWO, "name", pathToDisk2.getPath(), "path");
+        setUpDiskPool(disk1, disk2);
+        createUpstreamJobAndRun();
+
+        Disk allocatedDisk = findAllocatedDisk(disk1, disk2);
+
+        j.assertBuildStatusSuccess(upstreamRun);
+        j.assertLogContains(format("Selected Disk ID '%s' from the Disk Pool ID '%s'", allocatedDisk.getDiskId(), DISK_POOL_ID), upstreamRun);
+        j.assertLogContains(format("The path on Disk is: %s/%s/%d", allocatedDisk.getPhysicalPathOnDisk(), upstreamRun.getParent().getFullDisplayName(), upstreamRun.getNumber()), upstreamRun);
+    }
+
+    /* ##### Tests for the downstream Job ###### */
+
+    @Test
+    public void downstreamJobWithInvalidUpstreamName() throws Exception {
+        String upstreamName = "random-upstream-name";
+        createDownstreamJobAndRun(upstreamName);
+
+        j.assertBuildStatus(FAILURE, downstreamRun);
+        j.assertLogContains(format("Can't find any upstream Jenkins job by the full name '%s'. Are you sure that this is the full project name?", upstreamName), downstreamRun);
+    }
+
+    @Test
+    public void upstreamJobHasNoStableBuild() throws Exception {
+        createUpstreamJobAndRun("random-disk-pool");
+        String upstreamName = upstreamRun.getParent().getName();
+        createDownstreamJobAndRun(upstreamName);
+
+        j.assertBuildStatus(FAILURE, upstreamRun);
+        j.assertBuildStatus(FAILURE, downstreamRun);
+        j.assertLogContains(format("'%s' doesn't have any stable build", upstreamName), downstreamRun);
+    }
+
+    @Test
+    public void upstreamJobHasNoActionRegistered() throws Exception {
+        String jobName = randomAlphanumeric(10);
+        createWorkflowJobAndRun(jobName, "echo 'hello world'");
+        createDownstreamJobAndRun(jobName);
+
+        j.assertBuildStatus(FAILURE, downstreamRun);
+        j.assertLogContains(format("The Jenkins job '%s' does not have registered any 'External Workspace Allocate' action. Did you run exwsAllocate step in the upstream job?", jobName), downstreamRun);
```

Comment:
I think better grammar would be "... hasn't register any 'External ...", but I think the error message could be made more clear to the user.

When I write error messages in my code, I try to tell the user how they got in trouble in the first place, and how they can fix it. When I read this code, I ask myself: the lack of "run action" is a symptom of what? We want to make sure the novice user does not go on a tangent looking for ways to add a "run action", when all they have to do is call `exwsAllocate`. Ok, so let's park this idea and look at some corner cases.

I think there is a corner case where the user has a build history like this:
2. upstream calls `exwsAllocate` but fails **before** it registers the action.
1. upstream does not call `exwsAllocate` and is stable (green).

And another corner case, almost identical to the case above:
2. upstream calls `exwsAllocate` but fails **after** it registers the action
1. upstream does not call `exwsAllocate` and is stable (green)

We want the user to have a build that is stable and that calls `exwsAllocate`, so it has the "action" the code is looking for. Try to write an error message that guides the user in this direction.